sidh/sidh.go

```diff
@@ -0,0 +1,384 @@
+	var points = make([]ProjectivePoint, 0, 8)
+	var indices = make([]int, 0, 8)
+	var i, sidx int
+
+	cparam := curve.CalcCurveParamsEquiv3()
+	phi := NewIsogeny3()
+	strat := pub.params.B.IsogenyStrategy
+	stratSz := len(strat)
+
+	for j := 1; j <= stratSz; j++ {
+		for i <= stratSz-j {
+			points = append(points, *xR)
+			indices = append(indices, i)
+
+			k := strat[sidx]
+			sidx++
+			xR.Pow3k(&cparam, xR, k)
+			i += int(k)
+		}
+
+		cparam = phi.GenerateCurve(xR)
+		for k := 0; k < len(points); k++ {
+			points[k] = phi.EvaluatePoint(&points[k])
+		}
+
+		*phiP = phi.EvaluatePoint(phiP)
+		*phiQ = phi.EvaluatePoint(phiQ)
+		*phiR = phi.EvaluatePoint(phiR)
+
+		// pop xR from points
+		*xR, points = points[len(points)-1], points[:len(points)-1]
+		i, indices = int(indices[len(indices)-1]), indices[:len(indices)-1]
+	}
+}
+
+// Traverses isogeny tree in order to compute xR, xP, xQ and xQmP needed
+// for public key generation.
+func traverseTreeSharedKeyB(curve *ProjectiveCurveParameters, xR *ProjectivePoint, pub *PublicKey) {
+	var points = make([]ProjectivePoint, 0, 8)
+	var indices = make([]int, 0, 8)
+	var i, sidx int
+
+	cparam := curve.CalcCurveParamsEquiv3()
+	phi := NewIsogeny3()
+	strat := pub.params.B.IsogenyStrategy
+	stratSz := len(strat)
+
+	for j := 1; j <= stratSz; j++ {
+		for i <= stratSz-j {
+			points = append(points, *xR)
+			indices = append(indices, i)
+
+			k := strat[sidx]
+			sidx++
+			xR.Pow3k(&cparam, xR, k)
+			i += int(k)
+		}
+
+		cparam = phi.GenerateCurve(xR)
+		for k := 0; k < len(points); k++ {
+			points[k] = phi.EvaluatePoint(&points[k])
+		}
+
+		// pop xR from points
+		*xR, points = points[len(points)-1], points[:len(points)-1]
+		i, indices = int(indices[len(indices)-1]), indices[:len(indices)-1]
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Key generation functions
+//
+
+// Generate a private key for "Alice".  Note that because this library does not
+// implement SIDH validation, each keypair must be used for at most one
+// shared secret computation.
+func (prv *PrivateKey) generatePrivateKeyA(rand io.Reader) error {
+	_, err := io.ReadFull(rand, prv.Scalar)
+	if err != nil {
+		return err
+	}
+
+	// Bit-twiddle to ensure scalar is in 2*[0,2^371):
+	prv.Scalar[prv.params.SecretKeySize-1] = prv.params.A.MaskBytes[0]
+	prv.Scalar[prv.params.SecretKeySize-2] &= prv.params.A.MaskBytes[1] // clear high bits, so scalar < 2^372
+	prv.Scalar[0] &= prv.params.A.MaskBytes[2]                          // clear low bit, so scalar is even
+
+	// We actually want scalar in 2*(0,2^371), but the above procedure
+	// generates 0 with probability 2^(-371), which isn't worth checking
+	// for.
+	return nil
+}
+
+// Generate a private key for "Bob".  Note that because this library does not
+// implement SIDH validation, each keypair must be used for at most one
+// shared secret computation.
+func (prv *PrivateKey) generatePrivateKeyB(rand io.Reader) error {
```

Comment:
Why is this function so much more involved than for generating the key for A.

I am looking at https://github.com/Microsoft/PQCrypto-SIDH/blob/master/src/sidh.c#L60-L79 for reference and both of their functions seem fairly simple. Are they not taking as much care to sample a value with an even distribution

Reply:
That's good question. I don't know - code is copied from our old implementation, I would be guessing it tries to avoid leaking high bits. 
I haven't change it deliberately, so that amount of functionality changed is minimal. But it's going change anyway (quite soon), as this code is P751 specific and I need P503

Reply:
I think that's good idea. I actually have an alternative implementation that will be added at later stage.